src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinderAttribute.cs

```diff
@@ -26,39 +28,108 @@ namespace Microsoft.AspNetCore.Mvc
         Inherited = true)]
     public class ModelBinderAttribute : Attribute, IModelNameProvider, IBinderTypeProviderMetadata
     {
+        private static readonly BindingSource[] BindingSources = new[]
+        {
+            BindingSource.Body,
+            BindingSource.Custom,
+            BindingSource.Form,
+            BindingSource.FormFile,
+            BindingSource.Header,
+            BindingSource.ModelBinding,
+            BindingSource.Path,
+            BindingSource.Query,
+            BindingSource.Services,
+            BindingSource.Special,
+        };
+
         private BindingSource _bindingSource;
+        private Type _binderType;
 
         /// <summary>
         /// Initializes a new instance of <see cref="ModelBinderAttribute"/>.
         /// </summary>
+        /// <remarks>
+        /// If setting <see cref="BinderType"/> to an <see cref="IModelBinder"/> implementation that does not use values
+        /// from form data, route values or the query string, instead use the
+        /// <see cref="ModelBinderAttribute(BindingSourceKey)"/> constructor to set <see cref="BindingSource"/>.
+        /// </remarks>
         public ModelBinderAttribute()
         {
         }
 
+        /// <summary>
+        /// Initializes a new instance of <see cref="ModelBinderAttribute"/>.
+        /// </summary>
+        /// <param name="bindingSource">
+        /// The <see cref="BindingSourceKey"/> that indicates the value of the <see cref="BindingSource"/> property.
+        /// </param>
+        public ModelBinderAttribute(BindingSourceKey bindingSource)
+        {
+            var sourcesIndex = (int)bindingSource;
+            if (!Enum.IsDefined(typeof(BindingSourceKey), bindingSource))
+            {
+                throw new InvalidEnumArgumentException(nameof(bindingSource), sourcesIndex, typeof(BindingSourceKey));
+            }
+
+            _bindingSource = BindingSources[sourcesIndex];
+        }
+
         /// <summary>
         /// Initializes a new instance of <see cref="ModelBinderAttribute"/>.
         /// </summary>
         /// <param name="binderType">A <see cref="Type"/> which implements <see cref="IModelBinder"/>.</param>
+        /// <remarks>
+        /// If the specified <paramref name="binderType" /> does not use values from form data, route values or the
+        /// query string, instead use the <see cref="ModelBinderAttribute(BindingSourceKey)"/> constructor to set
+        /// <see cref="BindingSource"/> and set the <see cref="BinderType"/> property.
+        /// </remarks>
         public ModelBinderAttribute(Type binderType)
```

Comment:
I did _not_ add a `ModelBinderAttribute(Type binderType)` because this constructor already looks redundant to me because `BinderType` is settable. LMK if you want that added.